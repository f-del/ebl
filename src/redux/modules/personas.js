/***
 *    ██████╗ ███████╗██████╗ ██╗   ██╗ ██████╗███████╗██████╗ ███████╗
 *    ██╔══██╗██╔════╝██╔══██╗██║   ██║██╔════╝██╔════╝██╔══██╗██╔════╝
 *    ██████╔╝█████╗  ██║  ██║██║   ██║██║     █████╗  ██████╔╝███████╗
 *    ██╔══██╗██╔══╝  ██║  ██║██║   ██║██║     ██╔══╝  ██╔══██╗╚════██║
 *    ██║  ██║███████╗██████╔╝╚██████╔╝╚██████╗███████╗██║  ██║███████║
 *    ╚═╝  ╚═╝╚══════╝╚═════╝  ╚═════╝  ╚═════╝╚══════╝╚═╝  ╚═╝╚══════╝
 *
 */

const initialState = {
  list: {
    e6wTJIBsAw5puwwrNEsR: {
      Name: "Feature Team",
      Details: "Meta persona, representing all collaborators working with Paul",
      Needs: [
        "To understand WHAT the Product Owner wants to be able to design collaborativly the HOW",
        "To manage easly the list of things to do to finish the most quickly the User Story"
      ],
      Picture: ""
    },
    uOCF2GLFiJl6fvxdlq4b: {
      Details: "Product Owner for a new product, work in a startup",
      Name: "Paul Deligny",
      Needs: [
        "To have a clear understanding about the product vision & objectives",
        "To write real User Stories, ie have a Customer Driven thinking",
        "To share at the right time and with the rigth level of information the User Stories with his team",
        "To have a clear vision on the job he'll have to do AND the job already or currently done"
      ],
      Picture: ""
    }
  }
};

export default function(state = initialState, action) {
  return state;
}

/***
 *    ███████╗███████╗██╗     ███████╗ ██████╗████████╗ ██████╗ ██████╗ ███████╗
 *    ██╔════╝██╔════╝██║     ██╔════╝██╔════╝╚══██╔══╝██╔═══██╗██╔══██╗██╔════╝
 *    ███████╗█████╗  ██║     █████╗  ██║        ██║   ██║   ██║██████╔╝███████╗
 *    ╚════██║██╔══╝  ██║     ██╔══╝  ██║        ██║   ██║   ██║██╔══██╗╚════██║
 *    ███████║███████╗███████╗███████╗╚██████╗   ██║   ╚██████╔╝██║  ██║███████║
 *    ╚══════╝╚══════╝╚══════╝╚══════╝ ╚═════╝   ╚═╝    ╚═════╝ ╚═╝  ╚═╝╚══════╝
 *
 */
export function getAllPersonas(state) {
  return mapToArrayPersona(getPersonasList(state)) || [];
}

function getPersonasList(state) {
  return getPersonas(validateState(state)).list || {};
}

/* Functionnal helper */
function validateState(state) {
  if (state === undefined) throw new Error("State arg is mandatory");
  return state;
}
function getPersonas(state) {
  return state.personas || {};
}

function mapToArrayPersona(list) {
  return Array.from(Object.keys(list), key => ({
    ...getPersonaId(key),
    ...list[key]
  }));
}
function getPersonaId(key) {
  return {
    Id: key
  };
}
