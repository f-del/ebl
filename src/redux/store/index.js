import cards from "../modules/cards";
import { combineReducers } from "redux";

const rootReducer = combineReducers({ cards });
export default rootReducer;
