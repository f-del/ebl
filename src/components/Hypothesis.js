import React from "react";

import PropTypes from "prop-types";
import UiCard from "@material-ui/core/Card";
import UiCardContent from "@material-ui/core/CardContent";
import UiTypography from "@material-ui/core/Typography";
import Persona from "./Persona";
import CreateUserStoryCard from "../containers/CreateUserStoryCard";
import CardsBoard from "../containers/CardsBoard";
import { cardType, cardStatus } from "../redux/modules/cards";
import CardListByStatus from "../containers/CardsListByStatus";

function Hypothesis({ hypothesis, persona }) {
  function getNeedsIdx() {
    return hypothesis !== undefined ? hypothesis.Persona.NeedsIndex : undefined;
  }

  if (hypothesis !== undefined) return displayCard();
  else return null;

  function displayCard() {
    return (
      <React.Fragment>
        <UiCard>
          {persona && (
            <UiCardContent>
              <Persona
                persona={persona}
                addStory={false}
                selectedNeeds={getNeedsIdx()}
              />
            </UiCardContent>
          )}
        </UiCard>
        <UiCard>
          <UiCardContent>
            <UiTypography variant="h2">{hypothesis.Title}</UiTypography>
          </UiCardContent>
        </UiCard>
        <UiCard>
          <UiCardContent>
            <UiTypography variant="h3">
              Describe the {persona.Name} Journey
            </UiTypography>
            <CreateUserStoryCard hypothesisId={hypothesis.Id} />
            <CardsBoard type={cardType.UserStory}>
              <CardListByStatus
                type={cardType.UserStory}
                status={cardStatus.TODO}
              />
            </CardsBoard>
          </UiCardContent>
        </UiCard>
      </React.Fragment>
    );
  }
}

Hypothesis.propTypes = {
  hypothesis: PropTypes.object
};

export default Hypothesis;
