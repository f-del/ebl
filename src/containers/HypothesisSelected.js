import { connect } from "react-redux";
import { getPersona } from "../redux/modules/personas";
import Hypothesis from "../components/Hypothesis";
import { getHypothesisSelected } from "../redux/modules/ui";

const mapStateToProps = (state, ownsProp) => ({
  persona: getPersona(
    state,
    ownsProp.hypothesis !== undefined ? ownsProp.hypothesis.Persona.Id : ""
  ),
  hypothesis: getHypothesisSelected(state)
});

const mapDispatchToProps = (dispatch, ownsProp) => ({});

const HypothesisSelected = connect(
  mapStateToProps,
  mapDispatchToProps
)(Hypothesis);

export default HypothesisSelected;
