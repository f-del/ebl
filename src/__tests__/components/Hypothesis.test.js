import React from "react";
import { shallow, configure, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { createStore, applyMiddleware } from "redux";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import reducer from "../../redux/store/index";

import UiTypography from "@material-ui/core/Typography";

import Hypothesis from "../../components/Hypothesis";
import { entity_hypothesis_attached } from "../datas";
import personnasReducer from "../../redux/modules/personas";
import uiReducer from "../../redux/modules/ui";
import HypothesisSelected from "../../containers/HypothesisSelected";
import { entity_persona_created } from "../redux/personas";

configure({ adapter: new Adapter() });
var store, api;
describe("Componant tests", () => {
  beforeEach(() => {
    jest.resetAllMocks();
    api = {
      Cards: {
        Post: jest.fn(),
        Get: jest.fn()
      }
    };

    store = createStore(
      reducer,
      applyMiddleware(thunk.withExtraArgument({ api }))
    );
  });

  test("Not already loaded hypothesis", () => {
    const wrapper = shallow(<Hypothesis />);

    expect(wrapper.isEmptyRender()).toEqual(true);
  });

  test("Display an hypothesis, props hypothesis setted externaly", () => {
    const hypothesis = entity_hypothesis_attached();
    const wrapper = shallow(<Hypothesis hypothesis={hypothesis} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(UiTypography).length).toBe(1);
  });
});

describe("Containers tests", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });
  test("Expect props onAction is defined", () => {
    personnasReducer.getPersona = jest.fn().mockImplementationOnce(id => {
      return dispatch => {
        return entity_persona_created;
      };
    });
    uiReducer.getHypothesisSelected = jest.fn().mockImplementationOnce(id => {
      return dispatch => {
        return entity_persona_created;
      };
    });

    store = createStore(reducer, applyMiddleware(thunk));

    const wrapper = shallow(
      <Provider store={store}>
        <HypothesisSelected />,
      </Provider>
    );

    expect(wrapper.props("persona")).toBeDefined();
    expect(wrapper.props("hypothesis")).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
});
