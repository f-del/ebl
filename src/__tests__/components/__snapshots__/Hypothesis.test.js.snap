// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componant tests Display an hypothesis, props hypothesis setted externaly 1`] = `
<Fragment>
  <WithStyles(Card)>
    <WithStyles(CardContent)>
      <Persona
        addStory={false}
        persona={
          Object {
            "Details": "Product Owner for a new product, work in a startup",
            "Id": "uOCF2GLFiJl6fvxdlq4b",
            "Name": "Paul Deligny",
            "Needs": Array [
              "To have a clear understanding about the product vision & objectives",
              "To write real User Stories, ie have a Customer Driven thinking",
              "To share at the right time and with the rigth level of information the User Stories with his team",
              "To have a clear vision on the job he'll have to do AND the job already or currently done",
            ],
            "Picture": "",
          }
        }
        selectedNeeds={3}
      />
    </WithStyles(CardContent)>
  </WithStyles(Card)>
  <WithStyles(Card)>
    <WithStyles(CardContent)>
      <WithStyles(Typography)
        variant="h2"
      >
        test
      </WithStyles(Typography)>
    </WithStyles(CardContent)>
  </WithStyles(Card)>
  <WithStyles(Card)>
    <WithStyles(CardContent)>
      <WithStyles(Typography)
        variant="h3"
      >
        Describe the 
        Paul Deligny
         Journey
      </WithStyles(Typography)>
      <Connect(CreateCard)
        hypothesisId="mockedId"
      />
      <Connect(CardsBoard)
        type="USER-STORY"
      >
        <Connect(CardsList)
          status={Symbol(TODO)}
          type="USER-STORY"
        />
      </Connect(CardsBoard)>
    </WithStyles(CardContent)>
  </WithStyles(Card)>
</Fragment>
`;

exports[`Containers tests Expect props onAction is defined 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "storeState": Object {
        "cards": Object {
          "list": Array [],
          "status": Object {
            "HYPOTHESIS": Symbol(NULL),
            "TASK": Symbol(NULL),
            "USER-STORY": Symbol(NULL),
          },
        },
        "criterias": Object {
          "list": Object {
            "BASIC": Array [
              Object {
                "Id": "IdCritBas1",
                "Value": false,
              },
            ],
            "DEV": Array [
              Object {
                "Id": "IdCritDev1",
                "Text": "Development",
                "Value": false,
              },
              Object {
                "Id": "IdCritDev2",
                "Text": "Unit testing",
                "Value": false,
              },
            ],
          },
        },
        "personas": Object {
          "list": Object {
            "e6wTJIBsAw5puwwrNEsR": Object {
              "Details": "Meta persona, representing all collaborators working with Paul",
              "Name": "Feature Team",
              "Needs": Array [
                "To understand WHAT the Product Owner wants to be able to design collaborativly the HOW",
                "To manage easly the list of things to do to finish the most quickly the User Story",
              ],
              "Picture": "",
            },
            "uOCF2GLFiJl6fvxdlq4b": Object {
              "Details": "Product Owner for a new product, work in a startup",
              "Name": "Paul Deligny",
              "Needs": Array [
                "To have a clear understanding about the product vision & objectives",
                "To write real User Stories, ie have a Customer Driven thinking",
                "To share at the right time and with the rigth level of information the User Stories with his team",
                "To have a clear vision on the job he'll have to do AND the job already or currently done",
              ],
              "Picture": "",
            },
          },
        },
        "ui": Object {
          "currentHypothesis": undefined,
        },
      },
    }
  }
>
  <Connect(Hypothesis) />
  ,
</ContextProvider>
`;
